File: global/generators/python/generate_constants.py
#!/usr/bin/env python3
"""
Generate Python constants from YAML configuration
"""

import yaml
import json
from pathlib import Path
from typing import Dict, Any

def load_yaml_config() -> Dict[str, Any]:
    """Load the YAML configuration file."""
    config_path = Path(__file__).parent.parent.parent / "data" / "valheim-world.yml"
    with open(config_path, 'r') as f:
        return yaml.safe_load(f)

def load_validation_data() -> Dict[str, Any]:
    """Load validation data."""
    validation_path = Path(__file__).parent.parent.parent / "data" / "validation-data.yml"
    with open(validation_path, 'r') as f:
        return yaml.safe_load(f)

def load_rendering_config() -> Dict[str, Any]:
    """Load rendering configuration."""
    rendering_path = Path(__file__).parent.parent.parent / "data" / "rendering-config.yml"
    with open(rendering_path, 'r') as f:
        return yaml.safe_load(f)

def merge_biome_defaults(biomes: Dict[str, Any], defaults: Dict[str, Any]) -> Dict[str, Any]:
    """Merge biome defaults with individual biome properties."""
    merged = {}
    for name, biome in biomes.items():
        if name == "defaults":
            continue
        
        # Start with defaults
        merged_biome = defaults.copy()
        
        # Override with biome-specific properties
        merged_biome.update(biome)
        
        merged[name] = merged_biome
    
    return merged

def generate_python_constants():
    """Generate Python constants file."""
    config = load_yaml_config()
    validation = load_validation_data()
    rendering = load_rendering_config()
    
    # Extract world constants
    world = config["world"]
    coordinates = config["coordinates"]
    height = config["height"]
    
    # Merge biome defaults
    biomes = merge_biome_defaults(config["biomes"], config["biomes"]["defaults"])
    
    # Generate Python file
    output_path = Path(__file__).parent.parent.parent.parent / "backend" / "app" / "core" / "generated_constants.py"
    
    with open(output_path, 'w') as f:
        f.write('"""\n')
        f.write('Generated Valheim Constants\n')
        f.write('==========================\n')
        f.write('DO NOT EDIT - Generated from global/data/*.yml\n')
        f.write(f'Generated: {config["metadata"]["last_updated"]}\n')
        f.write('"""\n\n')
        
        # World constants
        f.write('# World dimensions\n')
        f.write(f'WORLD_RADIUS = {world["radius"]}\n')
        f.write(f'WORLD_DIAMETER = {world["diameter"]}\n')
        f.write(f'WORLD_BOUNDS = {world["bounds"]}\n')
        f.write(f'WATER_EDGE = {world["water_edge"]}\n\n')
        
        # Coordinate system
        f.write('# Coordinate system\n')
        f.write(f'COORDINATE_ORIGIN = {coordinates["origin"]}\n')
        f.write(f'COORDINATE_UNIT = "{coordinates["unit"]}"\n')
        f.write(f'COORDINATE_AXES = {coordinates["axes"]}\n\n')
        
        # Height system
        f.write('# Height system\n')
        f.write(f'SEA_LEVEL = {height["sea_level"]}\n')
        f.write(f'HEIGHT_MULTIPLIER = {height["multiplier"]}\n')
        f.write(f'HEIGHT_RANGE = {height["range"]}\n')
        f.write(f'OCEAN_THRESHOLD = {height["ocean_threshold"]}\n')
        f.write(f'MOUNTAIN_THRESHOLD = {height["mountain_threshold"]}\n\n')
        
        # Biome constants
        f.write('# Biome system\n')
        f.write('BIOMES = {\n')
        for name, biome in biomes.items():
            f.write(f'    "{name}": {{\n')
            f.write(f'        "id": {biome["id"]},\n')
            f.write(f'        "rgb": {biome["rgb"]},\n')
            f.write(f'        "hex": "{biome["hex"]}",\n')
            f.write(f'        "height_range": {biome["height_range"]},\n')
            f.write(f'        "distance_range": {biome["distance_range"]},\n')
            f.write(f'        "noise_threshold": {biome["noise_threshold"]},\n')
            f.write(f'        "polar_offset": {biome["polar_offset"]},\n')
            if biome["fallback_distance"] is not None:
                f.write(f'        "fallback_distance": {biome["fallback_distance"]},\n')
            if biome["min_mountain_distance"] is not None:
                f.write(f'        "min_mountain_distance": {biome["min_mountain_distance"]},\n')
            f.write('    },\n')
        f.write('}\n\n')
        
        # Convenience lookups
        f.write('# Convenience lookups\n')
        f.write('BIOME_IDS = {name: data["id"] for name, data in BIOMES.items()}\n')
        f.write('BIOME_COLORS_RGB = {data["id"]: data["rgb"] for data in BIOMES.values()}\n')
        f.write('BIOME_COLORS_HEX = {data["id"]: data["hex"] for data in BIOMES.values()}\n')
        f.write('BIOME_NAMES = {data["id"]: name for name, data in BIOMES.items()}\n\n')
        
        # Validation data
        f.write('# Validation data\n')
        f.write('VALIDATION_DATA = {\n')
        f.write(f'    "coordinate_ranges": {validation["coordinate_ranges"]},\n')
        f.write(f'    "height_ranges": {validation["height_ranges"]},\n')
        f.write(f'    "biome_distribution": {validation["biome_distribution"]},\n')
        f.write(f'    "performance": {validation["performance"]},\n')
        f.write('}\n\n')
        
        # Rendering config
        f.write('# Rendering configuration\n')
        f.write('RENDERING_CONFIG = {\n')
        f.write(f'    "resolutions": {rendering["resolutions"]},\n')
        f.write(f'    "rendering": {rendering["rendering"]},\n')
        f.write(f'    "shoreline": {rendering["shoreline"]},\n')
        f.write(f'    "filters": {rendering["filters"]},\n')
        f.write(f'    "visualization": {rendering["visualization"]},\n')
        f.write('}\n')

if __name__ == "__main__":
    generate_python_constants()
    print("Python constants generated successfully!")

File: global/generators/javascript/generate_constants.js
#!/usr/bin/env node
/**
 * Generate JavaScript constants from YAML configuration
 */

const fs = require('fs');
const yaml = require('js-yaml');
const path = require('path');

function loadYamlConfig() {
    const configPath = path.join(__dirname, '..', '..', 'data', 'valheim-world.yml');
    const yamlContent = fs.readFileSync(configPath, 'utf8');
    return yaml.load(yamlContent);
}

function loadValidationData() {
    const validationPath = path.join(__dirname, '..', '..', 'data', 'validation-data.yml');
    const yamlContent = fs.readFileSync(validationPath, 'utf8');
    return yaml.load(yamlContent);
}

function loadRenderingConfig() {
    const renderingPath = path.join(__dirname, '..', '..', 'data', 'rendering-config.yml');
    const yamlContent = fs.readFileSync(renderingPath, 'utf8');
    return yaml.load(yamlContent);
}

function mergeBiomeDefaults(biomes, defaults) {
    const merged = {};
    for (const [name, biome] of Object.entries(biomes)) {
        if (name === 'defaults') continue;
        
        // Start with defaults
        const mergedBiome = { ...defaults };
        
        // Override with biome-specific properties
        Object.assign(mergedBiome, biome);
        
        merged[name] = mergedBiome;
    }
    return merged;
}

function generateJavaScriptConstants() {
    const config = loadYamlConfig();
    const validation = loadValidationData();
    const rendering = loadRenderingConfig();
    
    // Extract constants
    const world = config.world;
    const coordinates = config.coordinates;
    const height = config.height;
    const biomes = mergeBiomeDefaults(config.biomes, config.biomes.defaults);
    
    // Generate JavaScript file
    const outputPath = path.join(__dirname, '..', '..', '..', 'procedural-export', 'client', 'valheim-constants.js');
    
    let content = '';
    
    // Header
    content += '// Generated Valheim Constants\n';
    content += '// ===========================\n';
    content += '// DO NOT EDIT - Generated from global/data/*.yml\n';
    content += `// Generated: ${config.metadata.last_updated}\n\n`;
    
    // World constants
    content += 'export const VALHEIM_WORLD = {\n';
    content += '    // World dimensions\n';
    content += `    RADIUS: ${world.radius},\n`;
    content += `    DIAMETER: ${world.diameter},\n`;
    content += `    WATER_EDGE: ${world.water_edge},\n\n`;
    
    // Coordinate system
    content += '    // Coordinate system\n';
    content += `    COORDINATE_ORIGIN: [${coordinates.origin[0]}, ${coordinates.origin[1]}],\n`;
    content += `    COORDINATE_UNIT: "${coordinates.unit}",\n`;
    content += '    COORDINATE_AXES: {\n';
    content += `        x: "${coordinates.axes.x}",\n`;
    content += `        y: "${coordinates.axes.y}",\n`;
    content += `        z: "${coordinates.axes.z}"\n`;
    content += '    },\n\n';
    
    // Height system
    content += '    // Height system\n';
    content += `    SEA_LEVEL: ${height.sea_level},\n`;
    content += `    HEIGHT_MULTIPLIER: ${height.multiplier},\n`;
    content += `    HEIGHT_RANGE: [${height.range[0]}, ${height.range[1]}],\n`;
    content += `    OCEAN_THRESHOLD: ${height.ocean_threshold},\n`;
    content += `    MOUNTAIN_THRESHOLD: ${height.mountain_threshold}\n`;
    content += '};\n\n';
    
    // Biome constants
    content += 'export const BIOMES = {\n';
    for (const [name, biome] of Object.entries(biomes)) {
        content += `    ${name}: {\n`;
        content += `        id: ${biome.id},\n`;
        content += `        rgb: [${biome.rgb[0]}, ${biome.rgb[1]}, ${biome.rgb[2]}],\n`;
        content += `        hex: "${biome.hex}",\n`;
        content += `        heightRange: [${biome.height_range[0]}, ${biome.height_range[1]}],\n`;
        content += `        distanceRange: [${biome.distance_range[0]}, ${biome.distance_range[1]}],\n`;
        content += `        noiseThreshold: ${biome.noise_threshold},\n`;
        content += `        polarOffset: ${biome.polar_offset},\n`;
        content += `        fallbackDistance: ${biome.fallback_distance},\n`;
        content += `        minMountainDistance: ${biome.min_mountain_distance}\n`;
        content += '    },\n';
    }
    content += '};\n\n';
    
    // Convenience lookups
    content += '// Convenience lookups\n';
    content += 'export const BIOME_IDS = Object.fromEntries(\n';
    content += '    Object.entries(BIOMES).map(([name, data]) => [name, data.id])\n';
    content += ');\n\n';
    
    content += 'export const BIOME_COLORS_RGB = Object.fromEntries(\n';
    content += '    Object.values(BIOMES).map(data => [data.id, data.rgb])\n';
    content += ');\n\n';
    
    content += 'export const BIOME_COLORS_HEX = Object.fromEntries(\n';
    content += '    Object.values(BIOMES).map(data => [data.id, data.hex])\n';
    content += ');\n\n';
    
    content += 'export const BIOME_NAMES = Object.fromEntries(\n';
    content += '    Object.entries(BIOMES).map(([name, data]) => [data.id, name])\n';
    content += ');\n\n';
    
    // Utility functions
    content += '// Utility functions\n';
    content += 'export function getBiomeName(biomeId) {\n';
    content += '    return BIOME_NAMES[biomeId] || `Unknown(${biomeId})`;\n';
    content += '}\n\n';
    
    content += 'export function getBiomeColorRgb(biomeId) {\n';
    content += '    return BIOME_COLORS_RGB[biomeId] || [255, 0, 255]; // Magenta for unknown\n';
    content += '}\n\n';
    
    content += 'export function getBiomeColorHex(biomeId) {\n';
    content += '    return BIOME_COLORS_HEX[biomeId] || "#FF00FF"; // Magenta for unknown\n';
    content += '}\n\n';
    
    content += 'export function isWithinWorldBounds(x, z) {\n';
    content += '    return Math.abs(x) <= VALHEIM_WORLD.RADIUS && Math.abs(z) <= VALHEIM_WORLD.RADIUS;\n';
    content += '}\n\n';
    
    content += 'export function distanceFromCenter(x, z) {\n';
    content += '    return Math.sqrt(x * x + z * z);\n';
    content += '}\n\n';
    
    content += 'export function worldToPixel(worldX, worldZ, resolution) {\n';
    content += '    const pixelX = Math.floor((worldX + VALHEIM_WORLD.RADIUS) * resolution / VALHEIM_WORLD.DIAMETER);\n';
    content += '    const pixelZ = Math.floor((worldZ + VALHEIM_WORLD.RADIUS) * resolution / VALHEIM_WORLD.DIAMETER);\n';
    content += '    return [pixelX, pixelZ];\n';
    content += '}\n\n';
    
    content += 'export function pixelToWorld(pixelX, pixelZ, resolution) {\n';
    content += '    const worldX = (pixelX * VALHEIM_WORLD.DIAMETER / resolution) - VALHEIM_WORLD.RADIUS;\n';
    content += '    const worldZ = (pixelZ * VALHEIM_WORLD.DIAMETER / resolution) - VALHEIM_WORLD.RADIUS;\n';
    content += '    return [worldX, worldZ];\n';
    content += '}\n';
    
    // Write file
    fs.writeFileSync(outputPath, content);
    console.log('JavaScript constants generated successfully!');
}

if (require.main === module) {
    generateJavaScriptConstants();
}

module.exports = { generateJavaScriptConstants };
